feat: containerd configuration for mirror registry (#292)
d2iq-labs/capi-runtime-extensions: docs/content/customization/generic/global-mirror.md

+++
title = "Global Image Registry Mirror"
+++

Add containerd image registry mirror configuration to all Nodes in the cluster.

When the `globalImageRegistryMirror` variable is set, `files` with configurations for [Containerd default mirror](https://github.com/containerd/containerd/blob/main/docs/hosts.md#setup-default-mirror-for-all-registries) will be added.

This customization will be available when the
[provider-specific cluster configuration patch]({{< ref "..">}}) is included in the `ClusterClass`.

## Example

To provide image registry mirror with CA certificate, specify the following configuration:

> If your registry mirror requires self signed CA certifate, create a Kubernetes Secret with keys for `ca.crt`:

Reviewer: ```suggestion
If your registry mirror requires a private or self-signed CA certificate, create a Kubernetes Secret with the `ca.crt` key populated with the CA certificate in PEM format:
```

Author: fixed: https://github.com/d2iq-labs/capi-runtime-extensions/pull/292/commits/c669b2bd08639e5ad40ba113dc775ddb8ff8770e

```shell
kubectl create secret generic my-mirror-ca-cert-secret \
  --from-file=ca.crt=registry-ca.crt
```

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: <NAME>
spec:
  topology:
    variables:
      - name: clusterConfig

>         value:

Reviewer 2: Can you please include here whatever is needed for `imageRegistries` for completeness. This example can include the `secretRef` and the ECR can just be the `url.

Author: fixed. 9014ba63f9e26c93683d932ce6df6ca77664ca8a

+++
title = "Global Image Registry Mirror"
+++

Add containerd image registry mirror configuration to all Nodes in the cluster.

When the `globalImageRegistryMirror` variable is set, `files` with configurations for [Containerd default mirror](https://github.com/containerd/containerd/blob/main/docs/hosts.md#setup-default-mirror-for-all-registries) will be added.

This customization will be available when the
[provider-specific cluster configuration patch]({{< ref "..">}}) is included in the `ClusterClass`.

## Example

To provide image registry mirror with CA certificate, specify the following configuration:

> If your registry mirror requires a private or self-signed CA certificate,

Reviewer 3: The API supports credential with username/password as well. Do we want to document an example for username/password, too?

Author: The API supports credentials for username and password for mirror, but it is not implemented for the globalImageRegistryMirror yet. user can provide mirror credentials in the `imageRegistries` block for now. 
Once we implement the support for mirror credentials explicitly, we can update this documentation.

Reviewer 3: Thanks, I missed that.

If I create a Secret with username/password and reference it, will I get an error? I would want that.

Author: Added https://github.com/d2iq-labs/capi-runtime-extensions/issues/329 to track the enhancement.

Reviewer 3: So, right now, a ref handles a Secret with a certificate, but not the username/password. That's fine. Thanks!

create a Kubernetes Secret with the `ca.crt` key populated with the CA certificate in PEM format:

```shell

> kubectl create secret generic my-mirror-ca-cert-secret \

Reviewer 3: nit -- we know it's a Secret :-) 
```suggestion
kubectl create secret generic my-mirror-ca-cert \
```

Author: https://github.com/d2iq-labs/capi-runtime-extensions/pull/292/commits/6e97e3e261f060c4a57836f5a64cc7af40fe219e

  --from-file=ca.crt=registry-ca.crt
```

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: <NAME>
spec:
  topology:
    variables:
      - name: clusterConfig
        value:
          globalImageRegistryMirror:
            url: https://my-mirror.io
            credentials:
              secretRef:

>                 name: my-mirror-ca-cert-secret

Reviewer 3: nit -- we know it's a Secret :-)
```suggestion
                name: my-mirror-ca-cert
```

Author: https://github.com/d2iq-labs/capi-runtime-extensions/pull/292/commits/6e97e3e261f060c4a57836f5a64cc7af40fe219e

+++
title = "Global Image Registry Mirror"
+++

Add containerd image registry mirror configuration to all Nodes in the cluster.

When the `globalImageRegistryMirror` variable is set, `files` with configurations for
[Containerd default mirror](https://github.com/containerd/containerd/blob/main/docs/hosts.md).

This customization will be available when the
[provider-specific cluster configuration patch]({{< ref "..">}}) is included in the `ClusterClass`.

## Example

> To provide image registry mirror with CA certificate, specify the following configuration:

Reviewer: ```suggestion
To provide an image registry mirror with a CA certificate, specify the following configuration:
```

> If your registry mirror requires a private or self-signed CA certificate,

Reviewer: nit
```suggestion
If the registry mirror requires a private or self-signed CA certificate,
```

create a Kubernetes Secret with the `ca.crt` key populated with the CA certificate in PEM format:

```shell
kubectl create secret generic my-mirror-ca-cert \
  --from-file=ca.crt=registry-ca.crt
```

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: <NAME>
spec:
  topology:
    variables:
      - name: clusterConfig
        value:
          globalImageRegistryMirror:
            url: https://my-mirror.io
            credentials:
              secretRef:
                name: my-mirror-ca-cert
```

Applying this configuration will result in following new files on the

> `KubeadmControlPlaneTemplate` and `KubeadmConfigTemplate`

Reviewer: ```suggestion
`KubeadmControlPlaneTemplate` and `KubeadmConfigTemplate` resources:
```

- `/etc/containerd/certs.d/_default/hosts.toml`
- `/etc/certs/mirror.pem`

> To use a public hosted image registry (ex. ECR) as mirror, specify the following configuration:

Reviewer: ```suggestion
To use a public hosted image registry (e.g. ECR) as a registry mirror, specify the following configuration:
```

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: <NAME>
spec:
  topology:
    variables:
      - name: clusterConfig
        value:
          globalImageRegistryMirror:
            url: https://123456789.dkr.ecr.us-east-1.amazonaws.com
```

Applying this configuration will result in following new files on the

> `KubeadmControlPlaneTemplate` and `KubeadmConfigTemplate`

Reviewer: ```suggestion
`KubeadmControlPlaneTemplate` and `KubeadmConfigTemplate` resources:
```
